packages/bsky/src/auth.ts

```diff
@@ -0,0 +1,85 @@
+import express from 'express'
+import * as ui8 from 'uint8arrays'
+import * as crypto from '@atproto/crypto'
+import { AuthRequiredError } from '@atproto/xrpc-server'
+import { DidResolver } from '@atproto/did-resolver'
+
+export const authVerifier =
+  (didResolver: DidResolver) =>
+  async (reqCtx: { req: express.Request; res: express.Response }) => {
+    const jwtStr = getJwtStrFromReq(reqCtx.req)
+    if (!jwtStr) {
+      throw new AuthRequiredError('missing jwt', 'MissingJwt')
+    }
+    const did = await verifyJwt(didResolver, jwtStr)
+    return { credentials: { did } }
+  }
+
+export const authOptionalVerifier =
+  (didResolver: DidResolver) =>
+  async (reqCtx: { req: express.Request; res: express.Response }) => {
+    if (!reqCtx.req.headers.authorization) {
+      return { credentials: { did: null } }
+    }
+    return authVerifier(didResolver)(reqCtx)
+  }
+
+const verifyJwt = async (
+  didResolver: DidResolver,
+  jwtStr: string,
+): Promise<string> => {
+  const parts = jwtStr.split('.')
+  if (parts.length !== 3) {
+    throw new AuthRequiredError('poorly formatted jwt', 'BadJwt')
+  }
+  const payload = parseB64UrlToJson(parts[1]) as JwtPayload
```

Comment:
We might consider verifying `payload` is an object.  Someone could craft a bad token where `payload` is `null`, and we'd end-up 500'ing when checking `exp` below.

Reply:
yup good call 👍